llama_index/llms/llm.py

```diff
@@ -314,3 +335,117 @@ def _extend_messages(self, messages: List[ChatMessage]) -> List[ChatMessage]:
                 *messages,
             ]
         return messages
+
+    def _as_query_component(self, **kwargs: Any) -> QueryComponent:
+        """Return query component."""
+        if self.metadata.is_chat_model:
+            return LLMChatComponent(llm=self)
+        else:
+            return LLMCompleteComponent(llm=self)
+
+
+class LLMCompleteComponent(QueryComponent):
+    """LLM completion component."""
+
+    llm: LLM = Field(..., description="LLM")
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def set_callback_manager(self, callback_manager: Any) -> None:
+        """Set callback manager."""
+        self.llm.callback_manager = callback_manager
+
+    def _validate_component_inputs(self, input: Dict[str, Any]) -> Dict[str, Any]:
+        """Validate component inputs during run_component."""
+        if "prompt" not in input:
+            raise ValueError("Prompt must be in input dict.")
+        # do special check to see if prompt is a list of chat messages
+        if isinstance(input["prompt"], get_args(List[ChatMessage])):
+            input["prompt"] = self.llm.messages_to_prompt(input["prompt"])
```

Comment:
ok, I know this is lowkey confusing, BUT, lets do this

```
if isinstance(input["prompt"], get_args(List[ChatMessage])):
    input["prompt"] = self.llm.messages_to_prompt(input["prompt"])
    input["prompt"] = validate_and_convert_stringable(input["prompt"])
else:
    input["prompt"] = validate_and_convert_stringable(input["prompt"])
    input["prompt"] = self.llm.completion_to_prompt(input["prompt"])
```

And then when calling `complete()`/`acomplete()`

```
response = self.llm.complete(prompt, formatted=True)
# or
response = await self.llm.acomplete(prompt, formatted=True)
```

Reply:
This will avoid double-formatting 😅 (llm.predict() is doing something similar)

Reply:
i can make the change, but yeah not exactly sure what's going on lol. do you have an example of completion_to_prompt?

Reply:
@jerryjliu 

```
def completion_to_prompt(completion):
  return f"[INST] {completion} [/INST] "
```

Reply:
the default is just a function that returns its input untouched

Reply:
oh so the main idea is to not run `completion_to_prompt` if it's a formatted list of chat messages?

Reply:
actually nvm i see, it's because `messages_to_prompt` already formats it

Reply:
You got it :)